File: packages/core-ts/src/storage/mode.ts
/**
 * Server mode management: local-only vs relay-configured.
 *
 * Persists user's choice to skip relay permanently by storing a mode marker
 * in the existing config.enc encrypted storage.
 */

import { deleteConfig, readConfig, writeConfig } from './config-file.js'

const MODE_KEY = '_mode'
const LOCAL_MODE_VALUE = 'local'

export type ServerMode = 'local' | 'configured' | null

/**
 * Mark server as local-only (user explicitly skipped relay).
 * Writes {"_mode": "local"} to config.enc[serverName].
 */
export async function setLocalMode(serverName: string): Promise<void> {
  await writeConfig(serverName, { [MODE_KEY]: LOCAL_MODE_VALUE })
}

/**
 * Get server mode.
 * Returns "local" if local mode set, "configured" if has other keys, null if empty.
 */
export async function getMode(serverName: string): Promise<ServerMode> {
  const config = await readConfig(serverName)
  if (config === null) {
    return null
  }

  if (config[MODE_KEY] === LOCAL_MODE_VALUE) {
    return 'local'
  }

  // Has keys other than _mode -> configured
  const nonModeKeys = Object.keys(config).filter((k) => k !== MODE_KEY)
  if (nonModeKeys.length > 0) {
    return 'configured'
  }

  return null
}

/**
 * Remove mode marker (allows relay to trigger again).
 * Deletes the server's config entry entirely.
 */
export async function clearMode(serverName: string): Promise<void> {
  await deleteConfig(serverName)
}

File: packages/core-py/src/mcp_relay_core/storage/mode.py
"""Server mode management: local-only vs relay-configured.

Persists user's choice to skip relay permanently by storing a mode marker
in the existing config.enc encrypted storage.
"""

import logging
from typing import Literal

from mcp_relay_core.storage.config_file import (
    delete_config,
    read_config,
    write_config,
)

logger = logging.getLogger(__name__)

_MODE_KEY = "_mode"
_LOCAL_MODE_VALUE = "local"

ServerMode = Literal["local", "configured"] | None


async def set_local_mode(server_name: str) -> None:
    """Mark server as local-only (user explicitly skipped relay).

    Writes {"_mode": "local"} to config.enc[server_name].

    Args:
        server_name: Server identifier.
    """
    await write_config(server_name, {_MODE_KEY: _LOCAL_MODE_VALUE})
    logger.debug("Set local mode for %s", server_name)


async def get_mode(server_name: str) -> ServerMode:
    """Get server mode.

    Args:
        server_name: Server identifier.

    Returns:
        "local" if local mode set, "configured" if has other keys, None if empty.
    """
    config = await read_config(server_name)
    if config is None:
        return None

    if config.get(_MODE_KEY) == _LOCAL_MODE_VALUE:
        return "local"

    # Has keys other than _mode (or _mode with different value) -> configured
    non_mode_keys = [k for k in config if k != _MODE_KEY]
    if non_mode_keys:
        return "configured"

    return None


async def clear_mode(server_name: str) -> None:
    """Remove mode marker (allows relay to trigger again).

    Deletes the server's config entry entirely.

    Args:
        server_name: Server identifier.
    """
    await delete_config(server_name)
    logger.debug("Cleared mode for %s", server_name)
